File: packages/swift/Sources/FrickSwift/FrickDraftStore.swift
import Foundation

/// Cross-device draft sync over the `MessageDraft` foundation object.
///
/// Mirrors the React `useDraft({ sync: true })` hook and the Kotlin
/// `FrickDraftStore`: writes go through `upsertObject` and reads come
/// from a single shared `subscribeObject("MessageDraft")` subscription
/// the actor manages internally. Per-conversation rows are addressed
/// by the cross-SDK id convention `"${userId}:${conversationId}"`,
/// so per-user isolation comes for free.
///
/// Conflict policy on the schema is `versionPrecondition` — two
/// devices typing at once race; the store handles the conflict by
/// re-issuing the write at the server-reported version (last-write-
/// wins). Two-device contention converges after a couple of round
/// trips.
public actor FrickDraftStore {
    public static func draftId(userId: String, conversationId: String) -> String {
        "\(userId):\(conversationId)"
    }

    private let socket: FrickSyncSocket
    private let userId: String

    private var subscribed = false
    private var versions: [String: Int] = [:]
    private var bodies: [String: String] = [:]
    private var continuations: [String: [UUID: AsyncStream<String>.Continuation]] = [:]
    private var listenerTask: Task<Void, Never>?

    public init(socket: FrickSyncSocket, userId: String) {
        self.socket = socket
        self.userId = userId
    }

    deinit {
        listenerTask?.cancel()
    }

    /// Returns the most recently observed draft body for the
    /// `(user, conversation)` pair, or `nil` if no row has yet
    /// arrived.
    public func currentDraft(for conversationId: String) -> String? {
        bodies[Self.draftId(userId: userId, conversationId: conversationId)]
    }

    /// Subscribe to changes for a single draft row. The stream
    /// yields the current body (if known) immediately, then every
    /// subsequent server-confirmed change. Multiple subscribers
    /// share a single underlying `subscribeObject("MessageDraft")`
    /// subscription.
    public func observe(conversationId: String) async -> AsyncStream<String> {
        await ensureSubscribed()
        let id = Self.draftId(userId: userId, conversationId: conversationId)
        let token = UUID()
        let stream = AsyncStream<String> { continuation in
            self.attach(id: id, token: token, continuation: continuation)
            if let existing = self.bodies[id] {
                continuation.yield(existing)
            }
            continuation.onTermination = { @Sendable [weak self] _ in
                Task { await self?.detach(id: id, token: token) }
            }
        }
        return stream
    }

    /// Persist `body` as the active draft. Idempotent — repeated
    /// calls with the same body collapse to one write per
    /// `versionPrecondition` round-trip.
    public func setDraft(_ body: String, for conversationId: String) async {
        let id = Self.draftId(userId: userId, conversationId: conversationId)
        let expected = versions[id]
        let payload: [String: Any] = [
            "userId": userId,
            "conversationId": conversationId,
            "body": body,
            "updatedAt": Int(Date().timeIntervalSince1970 * 1000),
        ]
        do {
            let version = try await socket.upsertObject(
                type: "MessageDraft",
                id: id,
                value: payload,
                expectedVersion: expected
            )
            versions[id] = version
            bodies[id] = body
        } catch {
            // Best-effort. The schema's `versionPrecondition` conflicts
            // arrive as a server error; the next setDraft call picks up
            // the corrected version from the inbound objectsDelta path.
        }
    }

    /// Clear the draft for a conversation. Emits an empty-body upsert
    /// since the foundation schema doesn't expose a delete primitive
    /// today.
    public func clearDraft(for conversationId: String) async {
        await setDraft("", for: conversationId)
    }

    // MARK: - Internal

    private func ensureSubscribed() async {
        if subscribed { return }
        subscribed = true
        listenerTask = Task { [weak self] in
            guard let self else { return }
            do {
                try await self.socket.subscribeObject(type: "MessageDraft")
                for try await event in await self.socket.events {
                    guard case let .objectsDelta(records, _) = event else { continue }
                    await self.ingest(records: records)
                }
            } catch {
                await self.markUnsubscribed()
            }
        }
    }

    private func ingest(records: [FrickObjectRecord]) {
        for record in records where record.type == "MessageDraft" {
            let body = record.value["body"] ?? ""
            bodies[record.id] = body
            for cont in continuations[record.id]?.values ?? [:].values {
                cont.yield(body)
            }
        }
    }

    private func attach(id: String, token: UUID, continuation: AsyncStream<String>.Continuation) {
        var entries = continuations[id] ?? [:]
        entries[token] = continuation
        continuations[id] = entries
    }

    private func detach(id: String, token: UUID) {
        guard var entries = continuations[id] else { return }
        entries.removeValue(forKey: token)
        if entries.isEmpty {
            continuations.removeValue(forKey: id)
        } else {
            continuations[id] = entries
        }
    }

    private func markUnsubscribed() {
        subscribed = false
    }
}

File: apps/android/frick/src/main/java/dev/frick/client/FrickDraftStore.kt
package dev.frick.client

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onStart
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Cross-device draft sync over the `MessageDraft` foundation object.
 * Mirrors the React `useDraft({ sync: true })` hook and the Swift
 * `FrickDraftStore`. Per-conversation rows are addressed by the
 * cross-SDK id convention `"${userId}:${conversationId}"`, so
 * per-user isolation comes for free without a policy hook.
 *
 * The schema's `versionPrecondition` merge policy means concurrent
 * typing on two devices races; this store surfaces the conflict via
 * the returned [Result] from [setDraft] and refreshes its tracked
 * version off the inbound `objectsDelta` so the next write picks the
 * server's view (last-write-wins by default).
 */
class FrickDraftStore(
    private val socket: FrickSyncSocket,
    private val userId: String,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
) {
    private val bodies = mutableMapOf<String, String>()
    private val versions = mutableMapOf<String, Int>()
    private val mutex = Mutex()
    private val emissions = MutableSharedFlow<Pair<String, String>>(extraBufferCapacity = 64)
    private var listenerJob: Job? = null
    private val subscribed = MutableStateFlow(false)

    companion object {
        fun draftId(userId: String, conversationId: String): String = "$userId:$conversationId"
    }

    /**
     * Returns a flow of the draft body for `(userId, conversationId)`,
     * starting with the current known body (or empty string if none
     * yet) and updating on every server-confirmed delta. The first
     * call also lazily kicks off the underlying object subscription.
     */
    fun flowDraft(conversationId: String): Flow<String> {
        ensureSubscribed()
        val id = draftId(userId, conversationId)
        return emissions.asSharedFlow()
            .filter { it.first == id }
            .map { it.second }
            .onStart {
                val initial: String = mutex.withLock { bodies[id].orEmpty() }
                emit(initial)
            }
    }

    /**
     * Upsert the draft body. Returns `Result.success(version)` on a
     * normal ack, `Result.failure(FrickHttpException)` on the
     * versionPrecondition conflict (the caller can decide to re-try
     * with the server's version, which the next call here will pick
     * up automatically once the inbound delta refreshes our cache).
     */
    suspend fun setDraft(conversationId: String, body: String): Result<Int> {
        val id = draftId(userId, conversationId)
        val expected = mutex.withLock { versions[id] }
        val payload = mapOf(
            "userId" to userId,
            "conversationId" to conversationId,
            "body" to body,
            "updatedAt" to System.currentTimeMillis(),
        )
        return try {
            val version = socket.upsertObject(
                type = "MessageDraft",
                id = id,
                value = payload,
                expectedVersion = expected,
            )
            mutex.withLock {
                versions[id] = version
                bodies[id] = body
            }
            emissions.tryEmit(id to body)
            Result.success(version)
        } catch (error: FrickHttpException) {
            Result.failure(error)
        }
    }

    /**
     * Clear the draft for a conversation. Emits an empty-body upsert
     * since the schema doesn't expose a delete primitive today.
     */
    suspend fun clearDraft(conversationId: String): Result<Int> = setDraft(conversationId, "")

    private fun ensureSubscribed() {
        if (subscribed.value) return
        subscribed.value = true
        listenerJob = scope.launch {
            try {
                socket.subscribeObject(type = "MessageDraft")
            } catch (_: Exception) {
                subscribed.value = false
                return@launch
            }
            socket.events.collect { event ->
                if (event !is FrickInboundEvent.Delta) return@collect
                for (record in event.objects) {
                    val type = record["type"] as? String ?: continue
                    if (type != "MessageDraft") continue
                    val id = record["id"] as? String ?: continue
                    @Suppress("UNCHECKED_CAST")
                    val value = record["value"] as? Map<String, Any?> ?: continue
                    val body = value["body"]?.toString().orEmpty()
                    mutex.withLock { bodies[id] = body }
                    emissions.tryEmit(id to body)
                }
            }
        }
    }
}
